### ISSUE TITLE
MSL: gl_Position is undetectable in tessellation stage IO using the SPIRV-Cross reflection API

### ISSUE THREAD

**Issue by Dredhog** [KhronosGroup/SPIRV-Cross#1639]
At Unity we've working on adding Metal tessellation support using SPIRV-Cross but have hit a limitation, hopefully there's already a way of doing this that we're unaware of.

The issue is that there does not seem to be a way to tell when and at what offset the generated MSL will contain a `gl_Position` member in IO interface for the Vertex or Hull kernels or the Domain vertex function as it's not reported along with the other `SPVC_RESOURCE_TYPE_STAGE_INPUT` or `SPVC_RESOURCE_TYPE_STAGE_OUTPUT` resource list members  even though it's in the generated MSL.

This results in the following problems when trying to validate and link up the different Metal stages:
1) To be able to set up control-point attribute data (`MTLVertexAttributeDescriptor` data in `MTLVertexDescriptor` ) for the post-tess vertex (domain) stage it is necessary to know the layout of the Hull kernel's control point output buffer. As the data transfer is done via buffer it does not seem possible to work around this and get the data layout from the Metal reflection for the Domain `MTLFunction`.
2) If we don't know that `gl_Position` is in the vertex kernel output or the hull kernel output, we cannot know the control point size/stride to be able to allocate minimally sized buffers for communication between the tessellation stages (this can be worked around by always assuming `gl_Position` is there and padding each point with  a `float4` though is suboptimal)
3) As we use `--msl-multi-patch-workgroup` the communication between the vertex and hull kernels happens directly through a buffer. We'd like to validate that the format is identical but since we cannot know if only one of them contains `gl_Position` this can't be done for all cases. Note that if `--msl-multi-patch-workgroup` is not used, stage IO validation is still not possible since Vertex stage output may or may not contain `gl_Position` even though it could be detected in the Hull stage input using the Metal runtime reflection (but not using SPIRV-Cross)
4) Similar as point 3, we'd like to be able to verify that the `gl_Position` post-tess vertex (domain) function input attribute was written to by the previous stage after compiling the two stages with SPIRV-Cross

Currently the only solution seems to be to manually parse out the `OpEntryPoint` instruction and to go over the list of `Interface <id>`s and only use SPIRV-Cross to check if any of them have a `BuiltIn` decoration of `Position`. This also requires us to assume that the `gl_Position` member will always be places at the end of the IO structs, as there's no way to get its offset in the IO buffers as well. As the placement looks like an arbitrary SPIRV-Cross implementation detail it does not seem like something we should rely on.

* The following is the source for the vert/hull/domain stages:
[tess_source.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189619/tess_source.txt)

* This is the associated SPIR-V after compiling with DXC (binaries downloaded from [shader playground](http://shader-playground.timjones.io/26226e408b07100cfc2d8351cf3d0184), `.o` files renamed to `.spv.txt` to upload to github)
[Vert.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189613/Vert.spv.txt)
[Hull.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189612/Hull.spv.txt)
[Domain.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189611/Domain.spv.txt)

* The following is the MSL generated with SPIRV-Cross
[Vert.msl.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189618/Vert.msl.txt) `./spirv-cross --msl --msl-vertex-for-tessellation Vert.spv.txt --output Vert.msl.txt`
[Hull.msl.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189617/Hull.msl.txt) `./spirv-cross --msl --msl-multi-patch-workgroup Hull.spv.txt --output Hull.msl.txt`
[Domain.msl.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6189616/Domain.msl.txt) `./spirv-cross --msl Domain.spv.txt --output Domain.msl.txt`

**Comment by HansKristian-Work** [comment]
@cdavis5e How is this handled in MoltenVK?

**Comment by cdavis5e** [comment]
We do reflection on the SPIR-V, not the MSL.

**Comment by HansKristian-Work** [comment]
@cdavis5e So, there is an API for add_msl_shader_input, which I believe was added to handle the input scenario where we want to match vertex <-> tess control scenario. It handles padding of vertex inputs, and emitting unused inputs to make sure the buffer layouts match up.

However, I cannot find a good way to deal with the output scenario. Like, in order to create a good input to tess control, we need to know the exact buffer layout of the vertex output, e.g.:

```
struct Vert_out
{
    float2 out_var_TEXCOORD0;
    float3 out_var_NORMAL;
    float4 out_var_INTERNALTESSPOS;
    float4 gl_Position;
};
```
in the vertex shader example above. I don't see an existing API to handle this, so I'm not sure how MoltenVK can deal with this.
We'd need to know some things:

- Which Output variable are actually emitted.
- In which order where output variables emitted.
- Which binary offsets (mostly relevant for tess eval input)
- How large is Vert_out (relevant for tess control and eval input)
- Which builtin variables are actually emitted.

**Comment by HansKristian-Work** [comment]
Ideally, we should probably add a `get_msl_shader_outputs` which mirrors add_msl_shader_input, so that we can use that to do cross stage linking somewhat easily.

**Comment by Dredhog** [comment]
>Ideally, we should probably add a `get_msl_shader_outputs` which mirrors add_msl_shader_input, so that we can use that to do cross stage linking somewhat easily.

That sounds like part of what we're looking, however `add_msl_shader_input` can't handle the case where inputs and oututs are not in the same order. To be able to detect a mismatch then, we'd still need a `get_msl_shader_inputs`, no? While considering this, I'd like to suggest a few additions to the general API:
1) Add the ability to change the order of MSL input struct members. This would allow to fix up the vertex<->tess control stage linkage in my example HLSL as the order of control stage inputs is different from the vertex stage outputs. To my understanding `add_msl_shader_input` can only change the format and the attribute number of existing members but not their relative order in the struct.
2) Add something like `remove_msl_shader_output` which would remove the stage output member from the vertex or hull output struct. This is useful as we can know at compile time if some vertex or tess control outputs won't be used by the next stage.

**Comment by HansKristian-Work** [comment]
>     2\. This is useful as we can know at compile time if some vertex or tess control outputs won't be used by the next stage.

It is invalid to use an input which is not generated by previous stage, so not sure how relevant this is.

> add_msl_shader_input can't handle the case where inputs and oututs are not in the same order. 

I think that would be guaranteed by the internal member sorting, but need to test this to make sure.

**Comment by Dredhog** [comment]
>It is invalid to use an input which is not generated by previous stage, so not sure how relevant this is.

Yes, but what I mean is that we can definitively know when the next stage will read an output using SPIRV-Cross. e.g. if we compile the tess stages in reverse order i.e. eval -> control -> vertex It is possible to use `is_msl_shader_input_used` and be sure when e.g. cetain tess eval inputs won't need to be written out by the tess control stage or when certain tess control inputs won't need to be written out by the vertex stage. This would save intermediate buffer memory at runtime and remove unnecessary calculations and writes (the actual Metal compiler would likely optimize out the calculations whose results are not written to any outputs).

>I think that would be guaranteed by the internal member sorting, but need to test this to make sure.

The Vert_out and Hull_in order in the .msl.txt files I've attached is unfortunately identical, but this was a lucky case as the original HLSL had identical sorting except for the SV_POSITION. In other cases the MSL will have members ordered as in the input SPIR-V/HLSL where the vert outputs can have different ordering than hull inputs. So the ordering can be an issue ATM. But if SPIRV-Cross did some robust sorting that would indeed likely solve the problem!

>As far as I understand, the final order of the vertex outputs is not really relevant, the important bit is that you can ensure that memory layout between the stages is deterministic, right? From what I understand, memory layout is messed up when:
>* An output (builtin or not) is generated by one stage, but not consumed by next (opposite case is invalid)
>* An output generates N vector components, input consumes M components, where M < N.

Yes, no objections here

**Comment by HansKristian-Work** [comment]
> Yes, but what I mean is that we can definitively know when the next stage will read an output using SPIRV-Cross. e.g. if we compile the tess stages in reverse order i.e. eval -> control -> vertex

I see, you basically want LTO here ... The MSL backend has no easy way to mask out outputs like that. Not impossible to add that, sure, but it's non-trivial.

**Comment by Dredhog** [comment]
While on the topic, I just noticed that `spvc_type_get_num_array_dimensions` and `spvc_type_array_dimension_is_literal` return incorrect values for some hull and domain shader `SPVC_RESOURCE_TYPE_STAGE_INPUT` or `SPVC_RESOURCE_TYPE_STAGE_OUTPUT` resource type IDs. The following code snippet:
```
unsigned  arrayLength = 1;
if((spvc_type_get_num_array_dimensions(resource.type_id) > 0) && spvc_type_array_dimension_is_literal(resource.type_id, 0))
    arrayLength = static_cast<unsigned>(spvc_type_get_array_dimension(resource.type_id, 0));
```
Results in `arrayLength` as marked in the following HLSL comments
```
struct HullInOut {
    float4 vertex : INTERNALTESSPOS;  // arrayLength = 3
    float3 normal : NORMAL;           // arrayLength = 3
};
```
```
struct HullInOut {
    float4 vertex : INTERNALTESSPOS;  // arrayLength = 3
    float3 normal[2] : NORMAL;        // arrayLength = 2
};
```
So for some reason, when no array dims are provided, `spvc_type_get_num_array_dimensions` returns 1 and  `spvc_type_array_dimension_is_literal(resource.type_id, 0)` returns `SVPC_TRUE`, in this case resulting in an array length of 3 which is incorrect

See the generated reflection yaml files from the `.spv.txt` files attached originally (these also have `array` of `3` and `array_size_is_literal` `true` :

[Hull.json.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6204666/Hull.json.txt) `./spirv-cross --msl ./Hull.spv.txt --reflect > Hull.json.txt`
[Domain.json.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6204664/Domain.json.txt) `./spirv-cross --msl ./Domain.spv.txt --reflect > Domain.json.txt`

It works correctly, however, on per patch constants generated by the Hull shader (see the `TANWEIGHTS` member)
```
struct PatchTessFactorAndConstantData {
    float edge [3] : SV_TessFactor;
    float inside : SV_InsideTessFactor;
    
    float3 vTangent[7]    : TANGENT;     // arrayLength = 7
    float2 vUV[5]         : TEXCOORD;    // arrayLength = 5
    float3 vTanUCorner[9] : TANUCORNER;  // arrayLength = 9
    float3 vTanVCorner[12] : TANVCORNER; // arrayLength = 12
    float4 vCWts          : TANWEIGHTS;  // arrayLength = 1
};
```
The results are also correct for vertex shader IO and Domain shader outputs

**Comment by HansKristian-Work** [comment]
>  float3 normal[2] : NORMAL;        // arrayLength = 2

I think you want to look at array dimension 1 (array_dimensions - 1) here. The innermost array type comes first, so that would be 2, then 3 for number of control points.

**Comment by HansKristian-Work** [comment]
@Dredhog I've got a rough idea now how to remove unused outputs at least. See the PR. It's still WIP, and next week is Easter, so it will take some time before I can complete this.

**Comment by Dredhog** [comment]
>The innermost array type comes first, so that would be 2, then 3 for number of control points.

Ah, thank you, that hadn't crossed my mind! With this there indeed appears to be enough context to get the correct arrayLength.

>@Dredhog I've got a rough idea now how to remove unused outputs at least.

Looks promising! A few points:
* I'd like to try it out, however we rely on the C API, so can't do it directly yet (mighty try adding it in locally myself, hopefully it won't be too hard)
* I'm assuming the masking API is expected to work both on Vert and Hull outputs? From testing I did yesterday I've noticed that only Hull outputs and Domain inputs are sorted (based on names with `gl_Position` ending up in the end), I see you've improved the sorting for buildins, **however**, Vert outputs and Hull inputs don't appear to be sorted and I don't see additional sorting calls added in this draft yet, is that something you plan on doing?
* With this masking API we'd also be able avoid to storing metadata on Hull output offsets with every shader! If with the masking API Hull output and Domain input can be made identical (disregarding gl_TessLevel* in Domain inputs) we can rebuild the offsets from the Metal `stageInputAttributes` of the post-tess vertex `MTLFunction`

**Comment by HansKristian-Work** [comment]
Added some CLI options in the latest PR and added some fixes, but masking out tess control point outputs is going to be very complicated ... The storage class changes from device -> threadgroup which complicates things, and rewriting control point builtins is even more annoying.

**Comment by Dredhog** [comment]
@HansKristian-Work I just now noticed that:
1. The `spvc_compiler_msl_is_shader_input_used` API when run on the Domain stage per-patch input attributes seems to always report the very first patch input constant (after sorting) as used, even when it isn't (in [this shader](http://shader-playground.timjones.io/21207416a3161b5febd52a343fae1215) the domain input `float3 consta : CONST_A` i.e. `in.var.CONST_A` is reported as used). Thought I would mention this bug here since you're currently looking on related code.
2. Unused domain stage control point input attributes **are not** stripped from the MSL by SPIRV-Cross. Sorry for not noticing this sooner, I had assumed this since Domain stage patch input constants **are** stripped.  This slightly reduces the utility of the masking API for the Hull <-> Domain case, but masking is still very useful for cases where different structs are used (as in my original example):
```
struct HullOut {
    float3 normal : NORMAL;
    float4 position : SV_POSITION;
    float4 vertex : INTERNALTESSPOS;
    float2 tex : TEXCOORD0;
}; 
```
```
struct DomainIn {
    float4 vertex : SV_POSITION;
    float3 normal : NORMAL;
};
```
Or for removing unused patch constants as SPIRV-Cross does strip these.

With the addition of the masking API maybe an option for SPIRV-Cross to automatically strip unused Domain stage input control point attributes from the MSL could be added as well?

**Comment by HansKristian-Work** [comment]
```
#version 450
layout(quads) in;

layout(location = 0) in vec4 vs[];

void main()
{
	gl_Position = vec4(1.0);
}
```

compiles to
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct main0_out
{
    float4 gl_Position [[position]];
};

[[ patch(quad, 0) ]] vertex main0_out main0()
{
    main0_out out = {};
    out.gl_Position = float4(1.0);
    return out;
}
```

**Comment by HansKristian-Work** [comment]
Well, the SPIR-V seems to be using IO blocks. SPIRV-Cross doesn't track usage based on block members, only OpVariable declarations. Ideally, DXC should not emit IO blocks here, just normal variables.

**Comment by Dredhog** [comment]
> Well, the SPIR-V seems to be using IO blocks. SPIRV-Cross doesn't track usage based on block members, only OpVariable declarations. Ideally, DXC should not emit IO blocks here, just normal variables.

Ok, makes sense that it depends on the SPIR-V

Curiously I tried the same with glslang with the following HLSL:
```
#define REPRODUCE_NO_STRIPPING 1

struct HullOut {
    float2 unuseda : UNUSED_A;
    float3 normal : USED_A;
    float4 unusedb : UNUSED_B;
    float4 vertex : USED_B;
};

[domain("tri")]
float4 Domain (const OutputPatch<HullOut,3> vi, float3 bary : SV_DomainLocation) : SV_POSITION
{
    float3 pp[3];
    
#if REPRODUCE_NO_STRIPPING
    for (int i = 0; i < 3; ++i)
        pp[i] = vi[i].normal + vi[i].vertex.xyz;
#else
    pp[0] = vi[0].normal + vi[0].vertex.xyz;
    pp[1] = vi[1].normal + vi[1].vertex.xyz;
    pp[2] = vi[2].normal + vi[2].vertex.xyz;
#endif
    
    return float4(pp[0]*bary.x + pp[1]*bary.y + pp[2]*bary.z, 1);
}

```
 and it behaves exactly like DXC (meaning `UNUSED_A` and `UNUSED_B` aren't stripped). ( glslang on [shader playground](http://shader-playground.timjones.io/8526b0b7cbafb7af5e2a49d03088b099) ) (dxc on [shader playground](http://shader-playground.timjones.io/a2eb578413e3b975d0352f859b05f845) ). Though it only reproduces when I access the control point data from a loop, if I unroll it, the members do get stripped. You can see this if you set the `REPRODUCE_NO_STRIPPING` to `0` in my example. So this is indeed very dependent on DXC's and glslang's codegen.

**Comment by HansKristian-Work** [comment]
Right ye. To resolve arrays of control points, DXC is likely loading the entire struct one by one into arrays, just so the shader can index into it, lots of false static use in that case.

**Comment by HansKristian-Work** [comment]
The current PR should be good, and I added C API support. I cannot justify spending more time on this for the time being.

**Comment by Dredhog** [comment]
>The current PR should be good, and I added C API support.

Great news! I'll test it out as soon as possible

It would still be nice if SPIRV-Cross had some API to query what the current stage inputs/outputs (primarily for builtins) so it would be obvious what to mask. Though I suppose we can parse this information out manually.

Small tangent, should I report the bug I mentioned in https://github.com/KhronosGroup/SPIRV-Cross/issues/1639#issuecomment-815013404 in a separate issue so it's not missed or has it already been addressed?

**Comment by Dredhog** [comment]
@HansKristian-Work I just tested it out a bit, and it seems it's not sorting vertex shader outputs and hull shader inputs, is there a reason for this?

**Comment by HansKristian-Work** [comment]
It should. I need an example.

**Comment by HansKristian-Work** [comment]
Vert out:
```
struct Vert_out
{
    float2 out_var_TEXCOORD0;
    float3 out_var_NORMAL;
    float4 out_var_INTERNALTESSPOS;
    float4 gl_Position;
};
```

```
struct Hull_in
{
    float2 in_var_TEXCOORD0;
    float3 in_var_NORMAL;
    float4 in_var_INTERNALTESSPOS;
    float4 gl_Position;
};
```
seems correct to me.

**Comment by Dredhog** [comment]
>I need an example
 
Source HLSL:
[vert_hull_no_sorting.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6287141/vert_hull_no_sorting.txt)

The SPIR-V (from DXC):
[vert.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6287145/vert.spv.txt)
[hull.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6287144/hull.spv.txt)

The MSL:
[vert.msl.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6287156/vert.msl.txt) ( `./spirv-cross --msl --msl-multi-patch-workgroup --msl-vertex-for-tessellation vert.spv.txt --output vert.msl.txt` )
[hull.msl.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6287155/hull.msl.txt) ( ` ./spirv-cross --msl --msl-multi-patch-workgroup --msl-vertex-for-tessellation hull.spv.txt --output hull.msl.txt` )

As you can see the results contain:
```
struct Vert_out
{
    float4 out_var_INTERNALTESSPOS;
    float3 out_var_NORMAL;
    float2 out_var_TEXCOORD0;
    float4 gl_Position;
};
```

```
struct Hull_in
{
    float2 in_var_TEXCOORD0;
    float3 in_var_NORMAL;
    float4 in_var_INTERNALTESSPOS;
    float4 gl_Position;
};
```

>As for the earlier comment, that doesn't look like a bug if DXC emits dummy loads

I mean the first part

> The spvc_compiler_msl_is_shader_input_used API when run on the Domain stage per-patch input attributes seems to always report the very first patch input constant (after sorting) as used, even when it isn't (in this shader the domain input float3 consta : CONST_A i.e. in.var.CONST_A is reported as used). Thought I would mention this bug here since you're currently working on related code.

I say this is a bug because here SPIRV-Cross does not generate the domain patch input in the MSL but says it is used.

**Comment by HansKristian-Work** [comment]
I think DXC screwed up location mapping, because:

Vertex output:
```
layout(location = 0) out vec4 out_var_INTERNALTESSPOS;
layout(location = 1) out vec3 out_var_NORMAL;
layout(location = 2) out vec2 out_var_TEXCOORD0;
```

Hull input:
```
layout(location = 0) in vec2 in_var_TEXCOORD0[];
layout(location = 1) in vec3 in_var_NORMAL[];
layout(location = 2) in vec4 in_var_INTERNALTESSPOS[];
```

**Comment by Dredhog** [comment]
>does not comply with D3D ordering requirements. User semantics must be emitted in correct order.

Oh, was not aware that that was the case. I assumed that this was allowed because SPIRV-Cross does do Hull <-> Domain sorting by name and would handle such a case so I didn't see a reason why Vert <-> Hull couldn't also be done.

**Comment by HansKristian-Work** [comment]
> does do Hull <-> Domain sorting by name

what?

**Comment by HansKristian-Work** [comment]
Ok, I see the bug with patch input. Builtin gl_Position finds the first vacant location 0, and it's mistakenly reported as "in-use".

**Comment by HansKristian-Work** [comment]
Should be fixed in PR.

**Comment by Dredhog** [comment]
>The locations are emitted in a funny order, and I suspect DXC decided to sort by name here, for whatever reason. Perhaps it's context sensitive and sorts by semantic name if no builtin semantics exist.

Sorry, I had somehow assumed that SPIRV-Cross was doing the sorting, my bad... glslang also appears to do this, possibly since it's based on DXC? The fact that our previous MSL compiler also did sorting for both Vert <-> Hull and Hull <-> Domain had me thinking this was required by D3D. Glad it's not relevant then 😄 

>Should be fixed in PR.

Perfect!

**Comment by HansKristian-Work** [comment]
> possibly since it's based on DXC

glslang came before DXC, but I suspect this is a common workaround for applications which get this rule wrong. SPIR-V only matches by Location decoration, and SPIRV-Cross sorts the interface structs by Location and then BuiltIn decorations.

**Comment by Dredhog** [comment]
I tested out the actual masking.

It works for most cases, however there's a bug where masking hull output arrays results in invalid MSL.

The dummy locals seems to be regular arrays:
```
    threadgroup float3 out_var_NORMAL[3][2];
    threadgroup float3 out_var_TANGENT[2];
```

whereas SPIRV-Cross tries to assign `spvUnsafeArray` constants to them:
```
constant spvUnsafeArray<float3, 2> _48 = spvUnsafeArray<float3, 2>({ float3(0.0), float3(0.0) });
constant spvUnsafeArray<float3, 4> _49 = spvUnsafeArray<float3, 4>({ float3(0.0), float3(0.0), float3(0.0), float3(0.0) });
```
```
        out_var_TANGENT = _48;
        patchOut.out_var_TANUCORNER = _49;
```

 resulting in a Metal compiler error.

`xcrun -sdk macosx metal -c hull_array_masking_test.metal -o MyLibrary.air`

```
hull_array_masking_test.metal:88:37: error: array type 'threadgroup float3 [2]' is not assignable
    out_var_NORMAL[gl_InvocationID] = param_var_v[gl_InvocationID].normal;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
hull_array_masking_test.metal:96:25: error: array type 'threadgroup float3 [2]' is not assignable
        out_var_TANGENT = _48;
        ~~~~~~~~~~~~~~~ ^
2 errors generated.
```

The source for the above example:
[hull_array_masking_test.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6289883/hull_array_masking_test.txt)

The SPIR-V from DXC:
[hull_array_masking_test.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6289884/hull_array_masking_test.spv.txt)

I maked out one control point array attribute and one patch constant array attribute with `./spirv-cross --msl --msl-multi-patch-workgroup --msl-vertex-for-tessellation --mask-stage-output-location 1 0 --mask-stage-output-location 3 0 hull_array_masking_test.spv.txt --output hull_array_masking_test.metal`

[hull_array_masking_test.metal.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6289891/hull_array_masking_test.metal.txt) (appended `.txt` for uplaod)

Small side note: masking all outputs leaves an empty struct, though the MSL compiler only seems to warn about the unused `gl_out` pointer so hopefully this is not an actual issue for Metal.

**Comment by Dredhog** [comment]
@HansKristian-Work I was testing if SPIRV-Cross maintains consistent sorting for builtins and found a bug with domain shader inputs in the linked PR.

SPIRV-Cross generates the following domain inputs (in `domain_builtin_sorting.spv.txt below`):
```
struct Domain_in
{
    float gl_CullDistance_0;
    float gl_CullDistance_1;
    float3 in_var_NORMAL [[attribute(0)]];
    float4 in_var_POSITION [[attribute(1)]];
    float4 gl_Position [[attribute(2)]];
    float gl_ClipDistance_0 [[attribute(0)]];
    float gl_ClipDistance_1 [[attribute(1)]];
};
```

Which is both sorted differently from the Hull output and has missing/shadowed attribute locations

For reference, here's the hull output (from `hull_builtin_sorting.metal.txt`):
```
struct Hull_out
{
    float3 out_var_NORMAL;
    float4 out_var_POSITION;
    float4 gl_Position;
    float gl_ClipDistance[2];
    float gl_CullDistance[2];
};
```

Both of these use the same control point structure in the HLSL (see `builtin_sorting.txt`)

This is the source HLSL:
[builtin_sorting.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6303353/builtin_sorting.txt)

The SPIR-V (from DXC):
[hull_builtn_sorting.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6303359/hull_builtn_sorting.spv.txt)
[domain_builtin_sorting.spv.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6303360/domain_builtin_sorting.spv.txt)

The MSL from SPIRV-Cross:
[hull_builtin_sorting.metal.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6303361/hull_builtin_sorting.metal.txt)
[domain_builtin_sorting.metal.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/6303363/domain_builtin_sorting.metal.txt)

**Comment by HansKristian-Work** [comment]
> It works for most cases, however there's a bug where masking hull output arrays results in invalid MSL.

Sigh ... yes. This is another edge case. I'm surprised you're using OpVariable initializers on control point arrays. Is this something DXC does by default?

```
struct Domain_in
{
    float gl_CullDistance_0;
    float gl_CullDistance_1;
    float3 in_var_NORMAL [[attribute(0)]];
    float4 in_var_POSITION [[attribute(1)]];
    float4 gl_Position [[attribute(2)]];
    float gl_ClipDistance_0 [[attribute(0)]];
    float gl_ClipDistance_1 [[attribute(1)]];
};
```
Looks pretty broken.

**Comment by Dredhog** [comment]
>Is this something DXC does by default?

Yes. The HLSL Hull shader is a basic passthrough:
```
// Hull per control point shader
[domain("tri")]
[partitioning("fractional_odd")]
[outputtopology("triangle_cw")]
[patchconstantfunc("HullConst")]
[outputcontrolpoints(3)]
HullInOut Hull (InputPatch<HullInOut,3> v, uint id : SV_OutputControlPointID)
{
    return v[id];
}
```
And the SPIR-V was downloaded directly from [shader playground](http://shader-playground.timjones.io/ea6fad68aa4e4604ef101b46e8e40a82) without modification so this is default DXC behaviour

**Comment by Dredhog** [comment]
Also noticed a small issue with the help menu text.

It currently says `[--mask-stage-builtin <Position|PointSize|ClipDistance|CullDistance>]:`
but only actually works with `--mask-stage-output-builtin`

**Comment by Dredhog** [comment]
Is there any way to tell what the array lengths of `gl_CullDistance` and `gl_ClipDistance` will be in stage inputs/ouputs?

`Compiler::has_active_builtin` seems insufficient for this

**Comment by HansKristian-Work** [comment]
Don't think so, no. Guess I'll need more reflection APIs to deal with that.

**Comment by HansKristian-Work** [comment]
Actually, you should be able to query it with existing API. If you iterate over inputs and outputs in get_resources(), you can query the types of said variables, and you can then detect builtins, their decoration and the array size.

Note that Position/PointSize/CullDistance/ClipDistance live in a gl_PerVertex Block, so you need to detect if a member has BuiltIn decoration set.

**Comment by Dredhog** [comment]
>Actually, you should be able to query it with existing API. If you iterate over inputs and outputs in get_resources(), you can query the types of said variables, and you can then detect builtins, their decoration and the array size.

Hmm... are you sure? I just retested it with the following:
```
spvc_result result = SPVC_SUCCESS;
spvc_set activeSet = {};
result = spvc_compiler_get_active_interface_variables(spvCompiler, &activeSet);

spvc_resources activeResources = {};
result = spvc_compiler_create_shader_resources(spvCompiler, &activeResources);
result = spvc_compiler_create_shader_resources_for_active_variables(spvCompiler, &activeResources, activeSet);

const spvc_reflected_resource* activeInputs = nullptr; size_t activeInputCount = 0;
result = spvc_resources_get_resource_list_for_type(activeResources, SPVC_RESOURCE_TYPE_STAGE_INPUT, &activeInputs, &activeInputCount);
const spvc_reflected_resource* activeOutputs = nullptr; size_t activeOutputCount = 0;
result = spvc_resources_get_resource_list_for_type(activeResources, SPVC_RESOURCE_TYPE_STAGE_OUTPUT, &activeOutputs, &activeOutputCount);

for(int i = 0; i < activeInputCount; i++)
{
    spvc_reflected_resource activeInput = activeInputs[i];
    const char* name = activeInput.name;
}

for(int i = 0; i < activeOutputCount; i++)
{
    spvc_reflected_resource activeOutput = activeOutputs[i];
    const char* name = activeOutput.name;
}
```
And I only see the names `in_var_SV_POSITION`, `in_var_SV_CullDistance`, `in_var_SV_ClipDistance`, `in_var_SV_CullDistance1` etc in the vertex input, where they treated not as builtins but as user input semantics.

The other stages don't don't have builtins in any of their inputs/outputs. Let me know if I misunderstood your explanation.

**Comment by HansKristian-Work** [comment]
You need to check if activeInput.id has a DecorationBuiltin with `spvc_compiler_has_decoration(activeInput.id, SpvDecorationBuiltIn) && spvc_compiler_get_decoration(activeInput.id, DecorationBuiltIn) == DecorationBuiltinClipDistance`. Or something like that.

**Comment by HansKristian-Work** [comment]
```
               OpDecorate %gl_ClipDistance BuiltIn ClipDistance
               OpDecorate %gl_CullDistance BuiltIn CullDistance
               OpDecorate %gl_Position BuiltIn Position
               OpDecorate %gl_Position_0 BuiltIn Position
               OpDecorate %in_var_NORMAL Location 0
               OpDecorate %in_var_POSITION Location 1
```

**Comment by Dredhog** [comment]
What I meant to say is that the builtins are not reported in the `activeInputs` and `activeOutputs` for hull and domain stages even though the following are all `SPVC_TRUE` for vert output, hull input/output and domain input/output
`spvc_compiler_has_active_builtin(spvCompiler, SpvBuiltInPosition, storageClass) `
`spvc_compiler_has_active_builtin(spvCompiler, SpvBuiltInClipDistance, storageClass)`
`spvc_compiler_has_active_builtin(spvCompiler, SpvBuiltInCullDistance, storageClass)`

**Comment by HansKristian-Work** [comment]
It seems like get_resource_list ignores any builtin variables. https://github.com/KhronosGroup/SPIRV-Cross/blob/master/spirv_cross.cpp#L856 Might break some apps to enable that though ...

**Comment by Dredhog** [comment]
I uploaded the sample I was using for the above test to ShaderPlayground so you could check out the SPIR-V:
[hull SPIR-V](http://shader-playground.timjones.io/5c88c11e295ecfc3143ddb8e82ff8b6a)
[domain SPIR-V](http://shader-playground.timjones.io/0afbf5c834b51321792d4065248f30bb)

Before I asked about SPIRV-Cross support for getting builtin array lengths I had looked into parsing this it out myself. I had already done it similar to how you described it but I ran into a dead end when trying to get the actual array lengths as DXC does it differently for hull and domain stages. In the above links you can see that it emits multiple `ClipDistance` and `CullDistance` builtin decorations for the Hull stage but only once for the Domain stage. So it was not obvious for me how to parse out the actual array length.

>Sounds like we will need another API to handle builtins.

That would be much appreciated!

**Comment by HansKristian-Work** [comment]
API for builtin stage IO reflection is in PR with C bindings.

**Comment by Dredhog** [comment]
Nice, I'll try to test it out today!

**Comment by HansKristian-Work** [comment]
The PR finally passes CI, so I'll let it sit there for now.

**Comment by Dredhog** [comment]
The new builtin query API appears to work well, I'm correctly getting array lengths in my tests!

However I can still reproduce the issue I mentioned [herre](https://github.com/KhronosGroup/SPIRV-Cross/issues/1639#issuecomment-817135824) if I take the attached SPIR-V (`hull_array_masking_test.spv.txt`) and run SPIRV-Cross and then the MSL compiler

Only change I see is that the line numbers have shifted slightly in the error message:
```
hull_array_masking_test.metal:90:37: error: array type 'threadgroup float3 [2]' is not assignable
    out_var_NORMAL[gl_InvocationID] = param_var_v[gl_InvocationID].normal;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
hull_array_masking_test.metal:98:25: error: array type 'threadgroup float3 [2]' is not assignable
        out_var_TANGENT = _48;
        ~~~~~~~~~~~~~~~ ^
2 errors generated.
```

**Comment by HansKristian-Work** [comment]
Latest PR should fix that issue.

**Comment by Dredhog** [comment]
Ok, just checked, it's indeed now fixed.

The PR looks good to go from my perspective, great work!
